website/versioned_docs/version-v1.x/circuits.md

```diff
@@ -7,45 +7,407 @@ sidebar_position: 7
+
+| Bits        | Value                      |
+| ----------- | -------------------------- |
+| 1st 53 bits | `0`                        |
+| 2nd 50 bits | `batchEndIndex`            |
+| 3rd 50 bits | `currentMessageBatchIndex` |
+| 4th 50 bits | `numSignUps`               |
+| 5th 50 bits | `maxVoteOptions`           |
+
+For instance, if `maxVoteOptions` is 25 and `batchEndIndex` is `5`, and all other values are 0, the following is the `packedVals` representation in hexadecimal:
+
+`140000000000000000000000000000000000019`
+
+##### `currentSbCommitment` and `newSbCommitment`
+
+The `currentSbCommitment` is the $poseidon_3$ hash of the state tree root, the ballot tree root, and a random salt. The purpose of the random salt, which should be unique to each batch, is to ensure that the value of `currentSbCommitment` always changes even if all the commands in a batch are invalid and therefore do not change the state tree or ballot tree root.
+
+The result of applying a batch of messages to `currentSbCommitment` is `newSbCommitment`.
+
+##### `currentSbSalt`
+
+The salt used to produce `currentSbCommitment` (see above).
+
+##### `newSbSalt`
+
+The salt used to produce `newSbCommitment` (see above).
+
+##### `msgSubrootPathElements`
+
+The index of each message in `msgs` is consecutive. As such, in order to prove that each message in `msgs` is indeed a leaf of the message tree, we compute the subtree root of `msgs`, and then verify that the subtree root is indeed a subroot of `msgRoot`.
+
+A simplified example using a tree of arity 2:
+
+```
+             r
+           /  \
+          s    ...
+       /    \
+      o     o
+     / \   / \
+   a   b  c  d
+```
+
+To prove that `a...d` are leaves of the tree with root `r`, we prove that the leaves have the subroot `s` with depth 2, and _then_ prove that `s` is a member of `r` at depth 1.
+
+The implementation for this is in the `QuinBatchLeavesExists` circuit in `https://github.com/privacy-scaling-explorations/maci/blob/dev/circuits/circom/trees/incrementalQuinTree.circom`.
+
+This method requires fewer circuit constraints than if we verified a Merkle proof for each leaf.
+
+#### Statements that the circuit proves
+
+1. That the prover knows the preimage to `inputHash` (see above)
+2. That the prover knows the preimage to `currentSbCommitment` (that is, the state root, ballot root, and `currentSbSalt`)
+3. That `maxVoteOptions <= (5 ^ voteOptionTreeDepth)`
+4. That `numSignUps <= (5 ^ stateTreeDepth)`
+5. That `coordPubKey` is correctly derived from `coordPrivKey`
+6. That `coordPubKey` is the preimage to the Poseidon hash of `coordPubKey` (provided by the contract)
+7. That each message in `msgs` exists in the message tree
+8. That after decrypting and applying each message, in reverse order, to the corresponding state and ballot leaves, the new state root, new ballot root, and `newSbSalt` are the preimage to `newSbCommitment`
+
+### Tally Votes
```

Comment:
Explanation would be nice here (but as I'm reviewing I'm thinking a lot of these documentation changes could be done in future PRs)